lib/IRGen/IRGenFunction.cpp

```diff
@@ -513,3 +513,249 @@ llvm::Value *IRGenFunction::alignUpToMaximumAlignment(llvm::Type *sizeTy, llvm::
+  } else {
+    Builder.CreateStore(Builder.CreateBitOrPointerCast(
+                            resultAddr.getAddress().getAddress(),
+                            contResultAddr->getType()->getPointerElementType()),
+                        Address(contResultAddr, pointerAlignment));
+  }
+
+  // Fill the current task (i.e the continuation) with the continuation
+  // information.
+  // currTask->ResumeTask = @llvm.coro.async.resume();
+  assert(currTask->getType() == IGM.SwiftTaskPtrTy);
+  auto currTaskResumeTaskAddr = Builder.CreateStructGEP(currTask,3);
+  auto coroResume =
+      Builder.CreateIntrinsicCall(llvm::Intrinsic::coro_async_resume, {});
+
+  assert(AsyncCoroutineCurrentResume == nullptr &&
+         "Don't support nested get_async_continuation");
+  AsyncCoroutineCurrentResume = coroResume;
+  Builder.CreateStore(
+      Builder.CreateBitOrPointerCast(coroResume, IGM.FunctionPtrTy),
+      Address(currTaskResumeTaskAddr, pointerAlignment));
+  // currTask->ResumeContext = &continuation_context;
+  auto currTaskResumeCtxtAddr = Builder.CreateStructGEP(currTask, 4);
+  Builder.CreateStore(
+      Builder.CreateBitOrPointerCast(continuationContext.getAddress(),
+                                     IGM.SwiftContextPtrTy),
+      Address(currTaskResumeCtxtAddr, pointerAlignment));
+
+  // Publish all the writes.
+  // continuation_context.awaitSynchronization =(atomic release) nullptr;
+  auto contAwaitSyncAddr =
+      Builder.CreateStructGEP(continuationContext.getAddress(), 1);
+  auto null = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 0);
+  auto atomicStore =
+      Builder.CreateStore(null, Address(contAwaitSyncAddr, pointerAlignment));
+  atomicStore->setAtomic(llvm::AtomicOrdering::Release,
+                         llvm::SyncScope::System);
+  out.add(unsafeContinuation);
+}
+
+void IRGenFunction::emitAwaitAsyncContinuation(
+    SILType unsafeContinuationTy, bool isIndirectResult,
+    Explosion &outDirectResult, llvm::BasicBlock *&normalBB,
+    llvm::PHINode *&optionalErrorResult, llvm::BasicBlock *&optionalErrorBB) {
+  assert(AsyncCoroutineCurrentContinuationContext && "no active continuation");
+  auto pointerAlignment = IGM.getPointerAlignment();
+
+  // First check whether the await reached this point first. Meaning we still
+  // have to wait for the continuation result. If the await reaches first we
+  // abort the control flow here (resuming the continuation will execute the
+  // remaining control flow).
+  auto contAwaitSyncAddr =
+      Builder.CreateStructGEP(AsyncCoroutineCurrentContinuationContext, 1);
+  auto null = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 0);
+  auto one = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 1);
+  auto results = Builder.CreateAtomicCmpXchg(
+      contAwaitSyncAddr, null, one,
+      llvm::AtomicOrdering::AcquireRelease /*success ordering*/,
```

Comment:
The ordering needs to be swapped, we continue with control flow if the swap failed (i.e all information is available: both the await point was  reached and the continuation was resumed).